File: frontend/src/components/learning/t071-no-community-surface.test.ts
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { test } from "node:test";

/**
 * T071: D-046 absence over the Student-facing learning surface.
 *
 * The external Discord/Telegram Course community link left MVP scope and is deferred to S18. S5
 * renders none of it — and, just as importantly, renders no placeholder, no disabled control, and
 * no "coming soon" state, because a placeholder is the feature arriving early with the work left
 * undone.
 *
 * The backend halves — `backend/internal/learning/` and the S5 migrations — are asserted in the Go
 * suite, where mutation row 16 of the required-mutations table proves the detector is not vacuous.
 */

/**
 * The forbidden set. `community`, `discord`, and `telegram` are T071's three names verbatim; the
 * rest are the same capability renamed, which is how this decision would most plausibly erode.
 * Matching happens against text with every non-alphanumeric character removed, so `community_link`,
 * `communityUrl`, `CommunityLink`, `community-link`, and `community link` are one concept.
 */
const communityConcepts = [
  "community",
  "discord",
  "telegram",
  "whatsapp",
  "groupchat",
  "discussiongroup",
  "socialgroup",
  "sociallink",
  "joinourgroup",
  "comingsoon",
];

const normalize = (source: string): string => source.toLowerCase().replace(/[^a-z0-9]+/g, "");

function frontendRoot(): string {
  return process.cwd().endsWith("/frontend") ? process.cwd() : path.join(process.cwd(), "frontend");
}

/** Every production source file under one authoritative root. */
function productionFiles(relativeRoot: string): string[] {
  const root = path.join(frontendRoot(), relativeRoot);
  assert.ok(
    fs.existsSync(root) && fs.statSync(root).isDirectory(),
    `T071 scan root ${relativeRoot} is missing; the detector would pass vacuously`
  );
  const collected: string[] = [];
  const walk = (directory: string) => {
    for (const entry of fs.readdirSync(directory, { withFileTypes: true })) {
      const full = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        walk(full);
        continue;
      }
      // Production only: a test that names the concept in order to forbid it is not the concept.
      if (/\.(ts|tsx)$/.test(entry.name) && !/\.test\.(ts|tsx)$/.test(entry.name)) {
        collected.push(full);
      }
    }
  };
  walk(root);
  return collected.sort();
}

const scanRoots = ["src/app/[locale]/learn", "src/components/learning"];

test("the Student learning surface carries no deferred community concept", () => {
  let scanned = 0;
  for (const relativeRoot of scanRoots) {
    const files = productionFiles(relativeRoot);
    assert.ok(files.length > 0, `T071 scan root ${relativeRoot} matched no files; the detector would pass vacuously`);
    scanned += files.length;

    for (const file of files) {
      // String literals, JSX text, object keys, and route paths are all in scope — only the
      // delimiters are normalized away, never the content.
      const normalized = normalize(fs.readFileSync(file, "utf8"));
      const relative = path.relative(frontendRoot(), file);
      for (const concept of communityConcepts) {
        assert.ok(
          !normalized.includes(concept),
          `${relative} carries the deferred community concept "${concept}"; D-046 defers it to S18, ` +
            `so S5 renders no link, no placeholder, and no coming-soon state`
        );
      }
    }
  }
  assert.ok(scanned >= 2, `T071 frontend scan covered only ${scanned} files`);
});

test("no learning API type exposes a community field", () => {
  // The read models are where a community link would arrive if the backend ever served one, so the
  // client contract is asserted as well as the screens that render it.
  const client = fs.readFileSync(path.join(frontendRoot(), "src/lib/api/learning.ts"), "utf8");
  const normalized = normalize(client);
  for (const concept of communityConcepts) {
    assert.ok(
      !normalized.includes(concept),
      `the learning API client declares the deferred community concept "${concept}"`
    );
  }
});

test("the concept scan recognises every spelling a rename would use", () => {
  for (const spelling of [
    "community_link",
    "communityLink",
    "CommunityURL",
    "community-url",
    "community link",
    "COMMUNITY_LINK",
    "discordInvite",
    "telegram_group",
    "Join our WhatsApp group",
    "groupChat",
    "discussion-group",
    "social_group",
    "Coming soon",
    '<a href="https://discord.gg/x">Join</a>',
  ]) {
    const normalized = normalize(spelling);
    assert.ok(
      communityConcepts.some((concept) => normalized.includes(concept)),
      `the concept scan missed "${spelling}" (normalized "${normalized}")`
    );
  }

  // And it stays usable: ordinary learning vocabulary must not trip it.
  for (const benign of [
    "lessonIdentity",
    "reportContext",
    "entitlementEvaluator",
    "courseHome",
    "commit the transaction",
    "communication is not the word",
  ]) {
    const normalized = normalize(benign);
    for (const concept of communityConcepts) {
      assert.ok(!normalized.includes(concept), `the concept scan false-positived on "${benign}" via "${concept}"`);
    }
  }
});

File: backend/internal/learning/t071_no_community_surface_test.go
package learning

import (
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"testing"
)

// T071: D-046 absence over the production build.
//
// The external Discord/Telegram Course community link left MVP scope and is deferred to S18. No
// slice authors it, stores it, serves it, or renders it before launch — so the honest state of S5
// is that the concept is not present at all, not that it is present and hidden.
//
// That distinction is why this is an absence *detector* rather than an inspection. "We looked and
// there was nothing" decays the moment someone adds a column; a test that fails when the concept
// appears keeps the decision enforced. Mutation row 16 of the required-mutations table exists to
// prove this detector is not vacuous: adding `community_link` to the S5 migration must turn it red.
//
// The frontend halves of D-046 — `frontend/src/app/[locale]/learn/` and
// `frontend/src/components/learning/` — are asserted in the frontend suite, where those files live.

// communityConcepts is the forbidden set. `community`, `discord`, and `telegram` are the three T071
// names verbatim; the rest are the same capability under another name, because a rename is the
// obvious way this decision would erode. Each is matched case-insensitively against text with
// separators removed, so `community_link`, `communityUrl`, `CommunityLink`, `community-link`, and
// `community link` are one concept rather than five spellings.
var communityConcepts = []string{
	"community",
	"discord",
	"telegram",
	"whatsapp",
	"groupchat",
	"discussiongroup",
	"socialgroup",
	"sociallink",
	"joinourgroup",
	"comingsoon",
}

// separators are removed before matching so no casing or delimiter style escapes the scan. Letters
// and digits survive; everything else is dropped.
var separators = regexp.MustCompile(`[^a-z0-9]+`)

func normalizeForConceptScan(source string) string {
	return separators.ReplaceAllString(strings.ToLower(source), "")
}

// t071Root is one authoritative scan root and the file kinds it owns.
type t071Root struct {
	relative   string
	extensions []string
	// productionOnly drops `_test.go`: T071 is about the production build, and a test that names a
	// concept in order to forbid it must not be mistaken for the concept being present.
	productionOnly bool
}

func backendRoot(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot resolve the backend root")
	}
	return filepath.Clean(filepath.Join(filepath.Dir(file), "../.."))
}

// collectT071Files walks one root and returns the files it owns.
func collectT071Files(t *testing.T, base string, root t071Root) []string {
	t.Helper()
	full := filepath.Join(base, root.relative)
	info, err := os.Stat(full)
	if err != nil || !info.IsDir() {
		// A silently vanished root would make this test pass by scanning nothing.
		t.Fatalf("T071 scan root %s is missing; the detector would pass vacuously", root.relative)
	}
	var files []string
	err = filepath.Walk(full, func(path string, entry os.FileInfo, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.IsDir() {
			return nil
		}
		if root.productionOnly && strings.HasSuffix(path, "_test.go") {
			return nil
		}
		for _, extension := range root.extensions {
			if strings.HasSuffix(path, extension) {
				files = append(files, path)
				return nil
			}
		}
		return nil
	})
	if err != nil {
		t.Fatalf("walking %s: %v", root.relative, err)
	}
	sort.Strings(files)
	return files
}

// s5MigrationRoot covers both S5 migrations, up and down. Row 16 mutates one of these files, so the
// scan must reach the SQL itself rather than a schema summary of it.
func t071BackendRoots() []t071Root {
	return []t071Root{
		{relative: "internal/learning", extensions: []string{".go"}, productionOnly: true},
		{relative: "internal/db/migrations", extensions: []string{".sql"}},
	}
}

// TestNoCommunitySurfaceInProtectedLearningOrItsMigrations is T071's backend half.
func TestNoCommunitySurfaceInProtectedLearningOrItsMigrations(t *testing.T) {
	base := backendRoot(t)

	scanned := 0
	for _, root := range t071BackendRoots() {
		files := collectT071Files(t, base, root)
		if len(files) == 0 {
			t.Fatalf("T071 scan root %s matched no files; the detector would pass vacuously", root.relative)
		}
		scanned += len(files)

		for _, path := range files {
			content, err := os.ReadFile(path)
			if err != nil {
				t.Fatalf("reading %s: %v", path, err)
			}
			// String literals, SQL identifiers, object keys, and route paths are all in scope —
			// only the delimiters are normalized away, never the content.
			normalized := normalizeForConceptScan(string(content))
			relative, relErr := filepath.Rel(base, path)
			if relErr != nil {
				relative = path
			}
			for _, concept := range communityConcepts {
				if strings.Contains(normalized, concept) {
					t.Fatalf("%s carries the deferred community concept %q; D-046 defers it to S18, "+
						"so S5 stores, serves, and renders none of it", relative, concept)
				}
			}
		}
	}

	// The S5 migrations must actually be among the files scanned, or row 16 could not be detected.
	migrations := collectT071Files(t, base, t071Root{relative: "internal/db/migrations", extensions: []string{".sql"}})
	var sawS5 bool
	for _, path := range migrations {
		if strings.Contains(filepath.Base(path), "0014_protected_learning") {
			sawS5 = true
		}
	}
	if !sawS5 {
		t.Fatal("the S5 protected-learning migration was not scanned; mutation row 16 could not be detected")
	}
	if scanned < 2 {
		t.Fatalf("T071 scanned only %d files", scanned)
	}
	t.Logf("T071 backend scan covered %d production files across %d roots", scanned, len(t071BackendRoots()))
}

// TestCommunityConceptScanRecognisesEverySpelling proves the normalization actually collapses the
// casing and delimiter variants a rename would use — the detector's own non-vacuity, independent of
// the mutation run.
func TestCommunityConceptScanRecognisesEverySpelling(t *testing.T) {
	for _, spelling := range []string{
		"community_link",
		"communityLink",
		"CommunityURL",
		"community-url",
		"community link",
		"COMMUNITY_LINK",
		"ALTER TABLE courses ADD COLUMN community_link TEXT;",
		"discordInvite",
		"telegram_group",
		"Join our WhatsApp group",
		"groupChat",
		"discussion-group",
		"social_group",
		"Coming soon",
	} {
		normalized := normalizeForConceptScan(spelling)
		matched := false
		for _, concept := range communityConcepts {
			if strings.Contains(normalized, concept) {
				matched = true
				break
			}
		}
		if !matched {
			t.Fatalf("the concept scan missed %q (normalized %q)", spelling, normalized)
		}
	}

	// And it does not fire on ordinary text, so the detector stays usable.
	for _, benign := range []string{
		"lessonIdentity", "reportContext", "entitlementEvaluator",
		"CREATE TABLE content_reports", "commit the transaction", "communication is not the word",
	} {
		normalized := normalizeForConceptScan(benign)
		for _, concept := range communityConcepts {
			if strings.Contains(normalized, concept) {
				t.Fatalf("the concept scan false-positived on %q via %q", benign, concept)
			}
		}
	}
}
